[DI] Add `exclude` to `TaggedIterator` and `TaggedLocator`

ruudk opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.1 
| Bug fix?      |no
| New feature?  | yes
| Deprecations? |no 
| Tickets       | 
| License       | MIT
| Doc PR        | 

With this change, you can exclude one or more services from a tagged iterator
or locator.

It's common when working with Delegator or Chain services that call a list of services that
implement an interface, to also want to implement the interface on Delegator/Chain.

An example of this is in the Symfony HttpKernel component:

https://github.com/symfony/symfony/blob/0d6e859db236e37b8baa2b2bf4c8b7d14d151570/src/Symfony/Component/HttpKernel/CacheClearer/CacheClearerInterface.php#L19-L25
https://github.com/symfony/symfony/blob/0d6e859db236e37b8baa2b2bf4c8b7d14d151570/src/Symfony/Component/HttpKernel/CacheClearer/ChainCacheClearer.php#L21-L31

The `ChainCacheClearer` implements `CacheClearerInterface` and calls a list of `CacheClearerInterface`
clearers.

If that would have been configured with `#[AutoconfigureTag]` and `#[TaggedIterator]`, the
`ChainCacheClearer` would receive itself in the `iterable $clearers`.

With this change, it can exclude itself and rely fully on autowire/configure.

Another example:

```php
#[AutoconfigureTag]
interface ErrorTracker
{
    public function trackError(string $error): void;
}

final class SentryErrorTracker implement ErrorTracker
{
    public function trackError(string $error): void
    {
        echo "Send error to Sentry\n";
    }
}

final class NewRelicErrorTracker implement ErrorTracker
{
    public function trackError(string $error): void
    {
        echo "Send error to NewRelic\n";
    }
}

final class DelegatingErrorTracker implements ErrorTracker
{
    public function __construct(
        #[TaggedIterator(ErrorTracker::class, exclude: self::class)]
        private iterable $trackers
    ) {}

    public function trackError(string $error): void
    {
        foreach($this->trackers as $tracker) {
            $tracker->trackError($error);
        }
    }
}

// Alias ErrorTracker interface to DelegatingErrorTracker service

final class MyController
{
    public function __construct(private ErrorTracker $errorTracker) {}
    public function __invoke()
    {
        $this->errorTracker->trackError('Hello, World!');
    }
}
```

Without this change, the `DelegatingErrorTracker` would receive itself in the tagged
iterator.

ruudk commented:
The fabbot.io suggestion doesn't make any sense.

ruudk commented:
@nicolas-grekas Curious what you think about this approach.

fractalzombie commented:
@ruudk Hi, this is the first time I've seen such use of tags, I don't quite understand the rationality of this solution.

ruudk commented:
I think it's quiet common when working with Delegator or Chain services that call a list of services that implement an interface, to also want to implement the interface on Delegator/Chain.

An example of this is in the Symfony HttpKernel component:

https://github.com/symfony/symfony/blob/0d6e859db236e37b8baa2b2bf4c8b7d14d151570/src/Symfony/Component/HttpKernel/CacheClearer/CacheClearerInterface.php#L19-L25

https://github.com/symfony/symfony/blob/0d6e859db236e37b8baa2b2bf4c8b7d14d151570/src/Symfony/Component/HttpKernel/CacheClearer/ChainCacheClearer.php#L21-L31

The `ChainCacheClearer` implements `CacheClearerInterface` and calls a list of `CacheClearerInterface` clearers. 

If that would have been configured with `#[AutoconfigureTag]` and `#[TaggedIterator]`, the `ChainCacheClearer` would receive itself in the `iterable $clearers`.

With this PR, it can exclude itself and rely fully on autowire/configure.

Jeroeny commented:
Should it be added to this method too [DependencyInjection/Loader/Configurator/ContainerConfigurator.php#L142](https://github.com/symfony/symfony/blob/6.1/src/Symfony/Component/DependencyInjection/Loader/Configurator/ContainerConfigurator.php#L142) ?

ruudk commented:
@Jeroeny Good one, added it to the to-do. First I would like to know if this is something that _could_ be merged. Then I can improve/finalize the PR.

wouterj commented:
> The fabbot.io suggestion doesn't make any sense.

Yes, you can ignore the suggestion. Fabbot tries really hard to act like a human, but sometimes they just can't hide being a robot :)

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Argument/TaggedIteratorArgument.php:
```diff
@@ -23,15 +23,17 @@ class TaggedIteratorArgument extends IteratorArgument
+     * @param string|null       $defaultIndexMethod    The static method that should be called to get each service's key when their tag doesn't define the previous attribute
+     * @param bool              $needsIndexes          Whether indexes are required and should be generated when computing the map
+     * @param string|null       $defaultPriorityMethod The static method that should be called to get each service's priority when their tag doesn't define the "priority" attribute
+     * @param string|array|null $exclude               Services to exclude from the iterator
```
no need to accept `string|null` here imho, only `array` is fine

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Tests/Fixtures/AutoconfiguredInterface2.php:
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace Symfony\Component\DependencyInjection\Tests\Fixtures;
```
please add the licence header to all files

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Attribute/TaggedIterator.php:
```diff
@@ -19,6 +19,7 @@ public function __construct(
         public ?string $indexAttribute = null,
         public ?string $defaultIndexMethod = null,
         public ?string $defaultPriorityMethod = null,
+        public null|string|array $exclude = null,
```
no need to make this nullable, defaulting to `[]` would be work

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:
```diff
@@ -247,13 +247,13 @@ private function autowireMethod(\ReflectionFunctionAbstract $reflectionMethod, a
                     if (TaggedIterator::class === $attribute->getName()) {
                         $attribute = $attribute->newInstance();
-                        $arguments[$index] = new TaggedIteratorArgument($attribute->tag, $attribute->indexAttribute, $attribute->defaultIndexMethod, false, $attribute->defaultPriorityMethod);
+                        $arguments[$index] = new TaggedIteratorArgument($attribute->tag, $attribute->indexAttribute, $attribute->defaultIndexMethod, false, $attribute->defaultPriorityMethod, $attribute->exclude);
```
`(array) $attribute->exclude` to cope with my other comments

nicolas-grekas reviewed (commented):
Works for me, here are some notes. Thanks for submitting.

ruudk committed:
Add `exclude` to TaggedIterator and TaggedLocator

With this change, you can exclude one or more services from a tagged iterator
or locator.

It's common when working with Delegator or Chain services that call a list of services that
implement an interface, to also want to implement the interface on Delegator/Chain.

An example of this is in the Symfony HttpKernel component:

https://github.com/symfony/symfony/blob/0d6e859db236e37b8baa2b2bf4c8b7d14d151570/src/Symfony/Component/HttpKernel/CacheClearer/CacheClearerInterface.php#L19-L25
https://github.com/symfony/symfony/blob/0d6e859db236e37b8baa2b2bf4c8b7d14d151570/src/Symfony/Component/HttpKernel/CacheClearer/ChainCacheClearer.php#L21-L31

The `ChainCacheClearer` implements `CacheClearerInterface` and calls a list of `CacheClearerInterface`
clearers.

If that would have been configured with `#[AutoconfigureTag]` and `#[TaggedIterator]`, the
`ChainCacheClearer` would receive itself in the `iterable $clearers`.

With this change, it can exclude itself and rely fully on autowire/configure.

Another example:

```php
#[AutoconfigureTag]
interface ErrorTracker
{
    public function trackError(string $error): void;
}

final class SentryErrorTracker implement ErrorTracker
{
    public function trackError(string $error): void
    {
        echo "Send error to Sentry\n";
    }
}

final class NewRelicErrorTracker implement ErrorTracker
{
    public function trackError(string $error): void
    {
        echo "Send error to NewRelic\n";
    }
}

final class DelegatingErrorTracker implements ErrorTracker
{
    public function __construct(
        #[TaggedIterator(ErrorTracker::class, exclude: self::class)]
        private iterable $trackers
    ) {}

    public function trackError(string $error): void
    {
        foreach($this->trackers as $tracker) {
            $tracker->trackError($error);
        }
    }
}

// Alias ErrorTracker interface to DelegatingErrorTracker service

final class MyController
{
    public function __construct(private ErrorTracker $errorTracker) {}
    public function __invoke()
    {
        $this->errorTracker->trackError('Hello, World!');
    }
}
```

Without this change, the `DelegatingErrorTracker` would receive itself in the tagged
iterator.

ruudk commented:
@nicolas-grekas Thanks for your review, I applied your feedback and added more tests. The CI fails due to unrelated errors on `6.1` branch.

fabpot commented:
Thank you @ruudk.
